Aztec: Implements basic state machine for editor state

astralbodies opened the pull request:
Closes #6422 

This adds the Publish button (previously called Post) to the Aztec editor. Instead of using a series of hacky methods to determine the state of the button it extracts the logic into clearly defined units of state that is unit testable. Other variables that affect state of the editor are also in the state context.

This is partially based off of @frosty's PostEditorSaveAction enum and WPPostViewController extension but made to be more abstract.

The main things this solves are:
* Publish button title.
* Publish button state (enabled/disabled).
* What text is shown during the publish process (Publishing...).

What it can/should solve but isn't part of this PR:
* Is Post-Post shown?
* Is there a secondary publish button and what does it say? (publish now, save draft, etc)
* Anything else that's related to the state of the editor that shouldn't turn into a pile of crazy If statements.

Some other TODOs that aren't part of this PR:
* **Submit for Review** is a lot of text to fit on the button label. Maybe just Submit? How are we handling it in cases where the button text is just too long in other languages?
* Submit for Review is also not currently supported in the hybrid editor. May need to validate how this all works. Currently the Aztec editor supports it but we're not setting the Bool properly if the user has or doesn't have the permission to publish.
* Update the state context when the post is being published to disable the publish button.
* Add some more unit tests to make sure all state switches are accounted for.

## How to Test
Make sure the publish button shows the right text and is enabled in these scenarios:
1. New post, no content. (Publish, disabled)
2. New post, with content. (Publish, enabled)
3. New post, set to draft. (Save)
4. New post, future dated. (Schedule)
5. New post, past dated. (Publish)
6. Existing published post. (Update)
7. Existing scheduled post. (Update)
8. Existing draft post. (Save)

There are a bunch of other scenarios but you get the point.

Needs review: @frosty, @jleandroperez

jleandroperez commented on WordPress/Classes/ViewRelated/Post/AztecPostViewController.swift:
```diff
@@ -150,18 +156,46 @@ class AztecPostViewController: UIViewController {
+            originalPostStatus = PostStatus(rawValue: postStatus) ?? .draft
+        } else {
+            originalPostStatus = .draft
+        }
```
Nipticky // feel free to disregard!!

Potential candidate for a helper method:

```
enum PostStatus {
   static func fromPost(post: Post) -> PostStatus {
 //...
   }
}
```

frosty replied:
Could be a good idea – defaulting to `.draft`. Could also be used in `observeValue(forKey:...)`

jleandroperez commented on WordPress/Classes/ViewRelated/Post/AztecPostViewController.swift:
```diff
@@ -150,18 +156,46 @@ class AztecPostViewController: UIViewController {
+
         super.init(nibName: nil, bundle: nil)
+
+        addObservers(toPost: post)
```
I _think_ it's safe to remove this `addObservers`, since it's already being done in the **didSet**.

Unless i'm missing something!

astralbodies replied:
@jleandroperez `didSet` isn't called when an `init` method sets the properties before `super` is called. Once super is called then didSet is called every future time the property is set. Stabby!!

jleandroperez commented on WordPress/Classes/ViewRelated/Post/PostEditorState.swift:
```diff
@@ -0,0 +1,296 @@
+    fileprivate var isBeingPublished = false {
+        didSet {
+            if isBeingPublished != oldValue {
+                publishActionAllowed = hasContent && !isBeingPublished
```
Still thinking if this snippet deserves a method its own method, or not (refreshActionAllowedState). It's just a line... perhaps i'm overthinking this one!

jleandroperez commented on WordPress/Classes/ViewRelated/Post/PostEditorState.swift:
```diff
@@ -0,0 +1,296 @@
+    }
+
+    var publishButtonText: String {
+        return editorState.action().publishActionLabel
```
`action.publishActionLabel`

jleandroperez commented on WordPress/Classes/ViewRelated/Post/PostEditorState.swift:
```diff
@@ -0,0 +1,296 @@
+    }
+
+    var publishVerbText: String {
+        return editorState.action().publishingActionLabel
```
`action.publishingActionLabel` (nipticky, sorry!

jleandroperez commented on WordPress/Classes/ViewRelated/Post/PostEditorState.swift:
```diff
@@ -0,0 +1,296 @@
+    }
+
+    var isPostPostShown: Bool {
+        return editorState.action().isPostPostShown
```
`action.isPostPostShown` (sorry, nipticky++)

astralbodies replied:
Agreed - I'll turn action() into a protocol property.

jleandroperez commented:
Drafts are displaying the `Publish` Action (vs `Update` in the Hybrid editor):

1. In a testing blog, create a Draft
2. Enable Aztec
3. Edit the Post. The top right action will read `Publish`

Is this expected?

jleandroperez commented:
Glitch 2:

We've got an overlap between navigationBar Action Button, and the Aztec Title. We should either cap the Blog Title's Width, or indicate somehow else that we're in Aztec. (I'd vote for both perhaps!)

![simulator screen shot feb 1 2017 11 57 51](https://cloud.githubusercontent.com/assets/1195260/22504265/bbfc4bb0-e875-11e6-8e07-7c181a9ef7bc.png)

jleandroperez commented:
### Note 3:
The `post`'s setter observer is not refreshing the `PostEditorStateContext`. I suspect we should be reinstantiating the `PostEditorStateContext`'s instance, whenever the `Post` changes.

IE: From Revision to Original.

jleandroperez commented:
### Update on Note 1:
Problem there is, precisely, that the `didChangeAction` delegate (which is in charge of setting the Publish Button's Title) is not getting called, initially.

We could either mimic KVO's "Post Initial Value" flag, or just add that to the `refreshInterface` helper?

frosty commented:
A few notes from putting this through its paces:

* When writing a new post, the Publish button is only enabled when text is added to the _body_ of the post, not the title. On the hybrid editor, typing characters into the title field enables the button. We should probably do the same in Aztec? (you could save a post with only a title)
* If I set a new post to Scheduled (setting a publish date), the button changes to Schedule (as expected). If I change the post back to Published (by tapping Publish Immediately), the button title stays as Schedule. In fact, even if I change the post status to Draft the title is still Schedule. Not the case with the hybrid editor.
* I'm seeing the same as @jleandroperez – existing posts are all showing Publish for me. I tested with existing drafts, scheduled posts, and published posts. The button says Publish for all of them.
* The existing button says "Post" not "Publish". Are we changing this?

Taking a look through the code now... :)

frosty commented on WordPress/Classes/ViewRelated/Post/PostEditorState.swift:
```diff
@@ -0,0 +1,292 @@
+        }
+    }
+
+    var isPostPostShown: Bool {
```
Looks like this isn't used yet and I might be misunderstanding its use, but with the hybrid editor we also show post-post for scheduled posts.

astralbodies replied:
It's not currently used yet - will be in another PR.

frosty commented on WordPress/Classes/ViewRelated/Post/PostEditorState.swift:
```diff
@@ -0,0 +1,292 @@
+
+    fileprivate var hasContent = false {
+        didSet {
+            publishActionAllowed = hasContent && !isBeingPublished
```
I know it's only one line, but we could extract this to a private `updatePublishActionAllowed` func and call it from both `didSet`s?

astralbodies replied:
Yup I'll make that change. 👍

frosty commented on WordPress/Classes/ViewRelated/Post/PostEditorState.swift:
```diff
@@ -0,0 +1,292 @@
+    }
+
+    func updated(postStatus: PostStatus, context: PostEditorStateContext) -> PostEditorActionState {
+        return self
```
Presumably this is why I can't seem to get the button out of Schedule state once I've entered it?

frosty commented on WordPress/Classes/ViewRelated/Post/PostEditorState.swift:
```diff
@@ -0,0 +1,292 @@
+fileprivate extension PostEditorActionState {
+    func isFutureDated(_ date: Date) -> Bool {
+        let oneMinute: TimeInterval = 60.0
+        let dateOneMinuteFromNow = Date(timeInterval: oneMinute, since: Date())
```
I would instead use `Calendar.current.date(byAdding: .minute, value: 1, to: date)`

Or you could potentially use `Calendar.current.compare(Date(), to: date, toGranularity: .minute)` to compare them instead.

frosty commented on WordPress/Classes/ViewRelated/Post/AztecPostViewController.swift:
```diff
@@ -489,6 +527,43 @@ private extension AztecPostViewController {
+    }
+
+    internal func context(_ context: PostEditorStateContext, didChangeAction: PostEditorAction) {
+        publishButton.title = context.publishButtonText
```
In both of these delegate methods, is there a reason you're going via the context instead of using the passed in action / bool?

astralbodies commented on WordPress/WordPressTest/PostEditorStateTests.swift:
```diff
@@ -0,0 +1,149 @@
+    }
+
+    func testContextChangedNewPostToDraft() {
+        context = PostEditorStateContext(originalPostStatus: .draft, userCanPublish: true, delegate: self)
```
I fixed this with a check-in - new posts will pass in a nil originalPostStatus now since it's not on the server. I was trying to match logic in Calypso which makes every post a draft when creating it which is something we should eventually emulate. 😄

frosty replied:
I definitely agree we should start every post off as a draft!

astralbodies commented on WordPress/WordPressTest/PostEditorStateTests.swift:
```diff
@@ -0,0 +1,149 @@
+        XCTAssertEqual(PostEditorAction.submitForReview, context.action, "should return 'Submit for Review' if the post is a draft and user can't publish")
+    }
+
+
```
Added unit tests to cover these scenarios.

astralbodies commented:
Responding to @frosty's main thread comments:

A few notes from putting this through its paces:

> When writing a new post, the Publish button is only enabled when text is added to the body of the post, not the title. On the hybrid editor, typing characters into the title field enables the button. We should probably do the same in Aztec? (you could save a post with only a title)

Done. Had to add a target/action for the editing changed event on the textview to watch for this. The postTitle property on the post isn't updated until editing is done.

> If I set a new post to Scheduled (setting a publish date), the button changes to Schedule (as expected). If I change the post back to Published (by tapping Publish Immediately), the button title stays as Schedule. In fact, even if I change the post status to Draft the title is still Schedule. Not the case with the hybrid editor.

Fixed.

> I'm seeing the same as @jleandroperez – existing posts are all showing Publish for me. I tested with existing drafts, scheduled posts, and published posts. The button says Publish for all of them.

Fixed.

> The existing button says "Post" not "Publish". Are we changing this?

I'm forcing us to match what Calypso says. I think @mattmiklic agreed when I asked about it. 😄

astralbodies commented:
@jleandroperez Here are responses to your notes:

> Note 1:
> Drafts are displaying the Publish Action (vs Update in the Hybrid editor):

Fixed.

> Note 2:
> We've got an overlap between navigationBar Action Button, and the Aztec Title. We should either cap the Blog Title's Width, or indicate somehow else that we're in Aztec. (I'd vote for both perhaps!)

We talked about this and removing the navbar title makes the most sense. It's getting in the way and we'll mark the editor as being Aztec is some other way like slightly changing the title bar color or something. Will address in a different PR.

> Note 3:
> The post's setter observer is not refreshing the PostEditorStateContext. I suspect we should be re-instantiating the PostEditorStateContext's instance, whenever the Post changes.

For now I moved post revision creation to happen before the UI is set up. I think it makes sense we do the revision operation first.

> Note 4:
> Post's Status Context didChangeAction delegate method (which is in charge of setting the Publish Button's Title) is not getting called with the initial value.

I think letting the context emit the current state is a good idea. (There's a similar feature in RxSwift ... just sayin...) For now I'll leave it as there are a few things I want to add to the context (like the secondary publish button and post-post functionality) and I'll add this in.

astralbodies commented:
@jleandroperez @frosty I *think* this is ready for another pass.

mattmiklic commented:
> I'm forcing us to match what Calypso says. I think @mattmiklic agreed when I asked about it. 😄

💯. `Post` in the iOS editor is an outlier; both Calypso and wp-android both use `Publish`.

frosty commented:
@astralbodies Changes look good! 👍 for all the comments you've added. Fixes the issues I had, and the tests all pass.

:shipit: from me, after the conflicts are fixed!

jleandroperez commented:
Beautiful code @astralbodies, all looking good!!

:shipit: !!!

astralbodies commented:
Thank you both so much for the time to review!
